### PR TITLE
Test cases for FileUploader service save and clone method.

### PR DESCRIPTION
**Pull request by laxmisalunkhe**

### CODE REVIEW

File: FileUploaderGrailsPlugin.groovy [hunk 1]
```diff
@@ -50,7 +50,7 @@ This plugin now supports uploading files to CDN for rackspace & amazon.
         def amazoneSecret = grailsApplication.config.fileuploader.AmazonSecret
 
         if(amazonKey instanceof ConfigObject || amazoneSecret instanceof ConfigObject) {
-            log.info "No username or key configured for file uploader amazon CDN service."
+            println "No username or key configured for file uploader amazon CDN service."
```
Comment (SAgrawal14) [inline FileUploaderGrailsPlugin.groovy]: Revert this and configure the logger.

File: test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        } catch (Exception e) {
+            log.warn "Exception occured while copying test file", e
+        }
+        testFile = new File("./${fileName}")
+
+        return testFile
+    }
+
+    def setup() {
+        fileUploaderService = new FileUploaderService()
```
Comment (SAgrawal14) [inline test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy]: Delete all these manual code of instantiating service instance. Grails will do it for you by simply injecting it to test case as you have already done in at Line 16.

File: test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        fileUploaderService.grailsApplication = grailsApplication
+        fileUploaderService.messageSource = getI18n()
+        group = "logo" // FileUploader Plugin Configuration group
+    }
+
+    def cleanup() {
+    }
+
+    void "Test: saveFile() when empty file parameter passed."() {
```
Comment (SAgrawal14) [inline test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy]: Also, let's follow a standard way of writing a test description. Write something like this: **test saveFile method when empty file parameter is passed**.

File: test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+    void "Test: saveFile() method should throw exception for missing fileuploader configuration."() {
+        given:
+        FileUploaderServiceException configException = null
+        File testFile = getTestFile()
+
+        when:
+        try {
+            ufileInstance = fileUploaderService.saveFile(group, testFile)
+        } catch (Exception e) {
```
Comment (SAgrawal14) [inline test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy]: You don't have to add a catch block for testing a throwable exception. Read about `thrown()` method in Spock framework and use that instead.

Comment (SAgrawal14) [inline test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy]: Same goes for all instances.

File: test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+        fileUploaderService.grailsApplication = grailsApplication
+        fileUploaderService.messageSource = getI18n()
+        group = "logo" // FileUploader Plugin Configuration group
+    }
+
+    def cleanup() {
+    }
+
+    void "Test: saveFile() when empty file parameter passed."() {
+        given:
```
Comment (SAgrawal14) [inline test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy]: Each & every spock block should have a description text.

File: test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+        def file = null
+
+        when:
+        ufileInstance = fileUploaderService.saveFile(group, file)
+
+        then: "Method should return null value"
+        assert ufileInstance == null
```
Comment (SAgrawal14) [inline test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy]: Also, check that, no **UFile** instance has been created.

File: test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+    }
+
+    void "Test: saveFile() method should throw exception for missing fileuploader configuration."() {
```
Comment (SAgrawal14) [inline test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy]: **test saveFile method, when file group configuration does not exists**

File: test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+
+        then: "Method should throw exception for missing file uploader plugin configuration."
+        assert configException.message == "No config defined for group [${group}]. " +
+            "Please define one in your Config file."
+
+        cleanup: "Deleting test file for other test cases."
+        testFile?.delete()
+    }
+
+    void "Test: saveFile() method should throw exception for invalid file extention."() {
```
Comment (SAgrawal14) [inline test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy]: **test saveFile method, when a file with invalid extension is passed**

File: test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+        given: "Applying FileUploader configuration and creating file with unauthorized extension"
+        setupConfig()
+        FileUploaderServiceException unauthorizedExtension = null
+        File testFile = getTestFile("test-logo.tif")
+
+        when:
```
Comment (SAgrawal14) [inline test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy]: **service method is called for a file with invalid extension**

File: test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+            " Allowed extensions for this upload are " + 
+            "${grailsApplication.config.fileuploader.logo.allowedExtensions}"
+
+        assert unauthorizedExtension?.message == exceptionMessage
+
+        cleanup: "Deleting test file for other test cases."
+        testFile?.delete()
+    }
+
+    void "Test: saveFile() method should throw exception for Max file size."() {
```
Comment (SAgrawal14) [inline test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy]: Change similar test descriptions accordingly.

File: test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+
+        then: "Method should throw exception for Max file size."
+        String exceptionMessage = "Sent file is bigger than allowed. Max file size is 1 kb"
+        assert maxFileSizeExtension?.message == exceptionMessage
+
+        cleanup: "Deleting test file for other test cases."
+        testFile?.delete()
+    }
+
+    void "Test: saveFile() method should save file successfully."() {
```
Comment (SAgrawal14) [inline test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy]: **test saveFile method for valid file**

File: test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+        File testFile = getTestFile()
+
+        when:
+        try {
+            ufileInstance = fileUploaderService.saveFile(group, testFile)
+        } catch (Exception e) {
+            exception = e
+        }
+
+        then: "Method should create UFile Instance successfully."
```
Comment (SAgrawal14) [inline test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy]: **method should create UFile instance and file should me moved to configured location locally**

File: test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+        assert exception == null
+        assert ufileInstance
```
Comment (SAgrawal14) [inline test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy]: Well, assert statements are optional.

File: test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+        assert ufileInstance.id != null
+        assert ufileInstance.name == "test-logo"
+        assert ufileInstance.extension == "png"
+        assert ufileInstance.type == UFileType.LOCAL
+        assert ufileInstance.fileGroup == group
+        assert ufileInstance.provider == null
```
Comment (SAgrawal14) [inline test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy]: Assert for path also and verify that file actually exists in that location.

File: test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+
+        cleanup: "Deleting test file for other test cases."
+        testFile?.delete()
+    }
+
+    void "Test: cloneFile() when ufileInstance parameter not passed."() {
+        given:
+        def file = null
```
Comment (SAgrawal14) [inline test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy]: You don't have to define this. Just pass `null` in the method call.

File: test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+
+        when: "When empty uFile parameter passed"
+        ufileInstance = fileUploaderService.cloneFile(group, file)
+
+        then: "Method should return null value"
+        assert ufileInstance == null
+    }
+
+    void "Test: cloneFile() when ufileInstance parameter passed."() {
```
Comment (SAgrawal14) [inline test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy]: **test cloneFile method for valid UFile instance**

**Comment by SAgrawal14** [comment]
@laxmisalunkhe Test cases are good. Let's add some more test condition for `saveFile()` method like when custom file name is passed with and without extension, when commons multipart files are being saved and again both with & without extension.

File: test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy (hunk 3 above)
Comment (laxmisalunkhe) [inline test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy]: Updated Test cases title.

File: test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy (hunk 5 above)
Comment (laxmisalunkhe) [inline test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy]: Updated test cases description.

File: test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy (hunk 6 above)
Comment (laxmisalunkhe) [inline test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy]: Added check for no UFile instance creation.

File: test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy (hunk 7 above)
Comment (laxmisalunkhe) [inline test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy]: Updated test case title

File: FileUploaderGrailsPlugin.groovy (hunk 1 above)
Comment (laxmisalunkhe) [inline FileUploaderGrailsPlugin.groovy]: Configured logger.

File: test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy (hunk 2 above)
Comment (laxmisalunkhe) [inline test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy]: Removed this code.

File: test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy (hunk 4 above)
Comment (laxmisalunkhe) [inline test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy]: Removed try-catch block and used spock framework `thrown()` method.

File: test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy (hunk 8 above)
Comment (laxmisalunkhe) [inline test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy]: Updated title

File: test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy (hunk 9 above)
Comment (laxmisalunkhe) [inline test/integration/com/lucastex/grails/fileuploader/FileUploaderServiceSpec.groovy]: Updated title

**Comment by laxmisalunkhe** [comment]
Added More test cases and also updated existing test cases title and description fields.

**Comment by SAgrawal14** [comment]
Moved to internal repo.

**Pull request closed without merging**